metabrainz/admin/quickbooks/views.py

```diff
@@ -0,0 +1,378 @@
+                continue
+
+            # If the customer is invoiced in arrears and the last invoice was from the prior quarter -> current
+            if is_arrears and invoices[0]['begin'] == pq_start and invoices[0]['end'] == pq_end:
+                current.append(cust)
+                continue
+
+            # Check to see if this is an annual invoice
+            try:
+                end_dt = invoices[0]['end_dt']
+                begin_dt = invoices[0]['begin_dt']
+                delta = end_dt - begin_dt
+                if delta.days > 359 and delta.days < 366:
+                    cust['name'] += " (annual)"
+                    if time.mktime(end_dt.timetuple()) <= time.time():
+                        end_date = datetime.date(end_dt.year + 1, end_dt.month, end_dt.day).strftime("%m-%d-%Y")
+                        begin_date = datetime.date(begin_dt.year + 1, begin_dt.month, begin_dt.day).strftime("%m-%d-%Y")
+                        self.add_new_invoice(invoices[0], cust, begin_date, end_date, today, 12, price)
+                        ready_to_invoice.append(cust)
+                    else:
+                        current.append(cust)
+
+                    continue
+            except TypeError:
+                wtf.append(cust)
+                continue
+
+            # If the end date is after the curent date, then consider it current
+            if time.mktime(end_dt.timetuple()) > time.time():
+                current.append(cust)
+                continue
+
+            # Everyone else, WTF?
+            wtf.append(cust)
+
+        return render_template("quickbooks/index.html", ready=ready_to_invoice, wtf=wtf, current=current)
+
+
+    @expose('/', methods=['POST'])
+    def submit(self):
+        '''
+        Parse the form submission, load invoices, copy the invoices, update the pertinent details and then save new invoice.
+        '''
+
+        customer = 0
+        invoices = []
+        while True:
+            customer_id = request.form.get("customer_%d" % customer)
+            if not customer_id:
+                break
+
+            if not request.form.get("create_%d" % customer):
+                customer += 1
+                continue
+
+            begin_date = request.form.get("begin_%d" % customer)
+            end_date = request.form.get("end_%d" % customer)
+            base_invoice = request.form.get("base_invoice_%d" % customer)
+            qty = request.form.get("qty_%d" % customer)
+            price = request.form.get("price_%d" % customer)
+
+            invoices.append( { 'customer' : customer_id, 'begin' : begin_date, 'end' : end_date, 'base_invoice' : base_invoice, 'qty' : qty, 'price' : price })
+            customer += 1 
+
+        # setup the access nonsense again
+        access_token = session.get('access_token', None)
+        refresh_token = session.get('refresh_token', None)
+        realm = session.get('realm', None)
+
+        if not session['access_token']:
```

Comment:
this should probably be `if not access_token`, since we get it 4 lines above?

Reply:
what happens when we have an access token but no realm or refresh token? Is that possible?

Reply:
It isn't possible from what I know.